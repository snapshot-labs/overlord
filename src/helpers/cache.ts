const cache = new Map<string, any>();
export async function withCache<T>(key: string, fetcherFn: () => Promise<T>): Promise<T> {
  if (cache.has(key)) {
    return cache.get(key);
  }

  const result = await fetcherFn();
  cache.set(key, result);
  return result;
}
