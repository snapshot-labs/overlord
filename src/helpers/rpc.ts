import { Contract } from '@ethersproject/contracts';
import { StaticJsonRpcProvider } from '@ethersproject/providers';

export async function getTokenDecimals(network: number, address: string): Promise<number> {
  const url = `https://brovider.xyz/${network}`;
  const provider = new StaticJsonRpcProvider(url, network);
  const contract = new Contract(address, ['function decimals() view returns (uint8)'], provider);

  return await contract.decimals();
}
