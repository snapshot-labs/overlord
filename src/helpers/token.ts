import { Contract } from '@ethersproject/contracts';
import { StaticJsonRpcProvider } from '@ethersproject/providers';

const cache = new Map<string, number>();

export async function getTokenDecimals(network: number, address: string): Promise<number> {
  const key = `${network}:${address}`;

  if (cache.has(key)) return cache.get(key)!;

  const url = `https://brovider.xyz/${network}`;
  const provider = new StaticJsonRpcProvider(url, network);
  const contract = new Contract(address, ['function decimals() view returns (uint8)'], provider);

  const decimals = await contract.decimals();

  cache.set(key, decimals);

  return decimals;
}
