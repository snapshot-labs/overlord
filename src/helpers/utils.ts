import { Response } from 'express';

const HTTP_STATUS_MESSAGES: Record<number, string> = {
  400: 'Bad Request',
  500: 'Internal Server Error'
};

export function rpcSuccess(res: Response, result: any, id: number) {
  res.json({
    jsonrpc: '2.0',
    result,
    id
  });
}

export function rpcError(res: Response, code: number, e: unknown, id: number) {
  const message = HTTP_STATUS_MESSAGES[code] || 'unauthorized';

  res.status(code).json({
    jsonrpc: '2.0',
    error: {
      code,
      message,
      data: e
    },
    id
  });
}
