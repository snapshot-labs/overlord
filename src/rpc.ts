import express from 'express';
import { rpcError, rpcSuccess } from './helpers/utils';
import { validateRequest } from './middleware/validation';
import getStrategiesValue from './strategies';

const router = express.Router();

router.post('/', validateRequest, async function (req, res) {
  const {
    params: { network, snapshot, strategies },
    id = 0
  } = req.body;

  try {
    return rpcSuccess(res, await getStrategiesValue(network, snapshot, strategies), id);
  } catch (err) {
    return rpcError(res, 500, err, id);
  }
});

export default router;
