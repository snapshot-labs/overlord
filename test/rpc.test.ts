// @ts-ignore
import { describe, expect, it } from 'bun:test';
import { getTokenDecimals } from '../src/helpers/rpc';

describe('getTokenDecimals', () => {
  it('should get token decimals for UNI token on Ethereum', async () => {
    const result = await getTokenDecimals(
      1, // Ethereum mainnet
      '0x1f9840a85d5af5bf1d1762f925bdaddc4201f984' // UNI token address
    );
    expect(result).toMatchSnapshot();
  });

  it('should get token decimals for USDC token on Base', async () => {
    const result = await getTokenDecimals(
      8453, // Base network
      '0x833589fcd6edb6e08f4c7c32d4f71b54bda02913' // USDC token address
    );
    expect(result).toMatchSnapshot();
  });
});
