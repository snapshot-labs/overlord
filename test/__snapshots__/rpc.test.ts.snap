// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`getTokenDecimals should get token decimals for UNI token on Ethereum 1`] = `18`;

exports[`getTokenDecimals should get token decimals for USDC token on Base 1`] = `6`;
